chronos/apps/http/httpclient.nim

```diff
@@ -284,29 +298,75 @@ template checkClosed(reqresp: untyped): untyped =
     reqresp.setError(e)
     raise e
 
+template setTimestamp(conn: HttpClientConnectionRef,
+                      moment: Moment): untyped =
+  if not(isNil(conn)):
+    conn.timestamp = moment
+
+template setTimestamp(
+           reqresp: HttpClientRequestRef|HttpClientRequestRef
+         ): untyped =
+  if not(isNil(reqresp)):
+    let timestamp = Moment.now()
+    reqresp.timestamp = timestamp
+    reqresp.connection.setTimestamp(timestamp)
+
+template setTimestamp(resp: HttpClientResponseRef, moment: Moment): untyped =
+  if not(isNil(resp)):
+    resp.timestamp = moment
+    resp.connection.setTimestamp(moment)
+
+template setDuration(
+           reqresp: HttpClientRequestRef|HttpClientResponseRef
+         ): untyped =
+  if not(isNil(reqresp)):
+    let timestamp = Moment.now()
+    reqresp.duration = timestamp - reqresp.timestamp
+    reqresp.connection.setTimestamp(timestamp)
+
+template isReady(conn: HttpClientConnectionRef): bool =
+  (conn.state == HttpClientConnectionState.Ready) and
+  (HttpClientConnectionFlag.KeepAlive in conn.flags) and
+  (HttpClientConnectionFlag.Request notin conn.flags) and
+  (HttpClientConnectionFlag.Response notin conn.flags)
+
+template isIdle(conn: HttpClientConnectionRef, timestamp: Moment,
+                timeout: Duration): bool =
+  (timestamp - conn.timestamp) >= timeout
+
+proc sessionWatcher(session: HttpSessionRef) {.async.}
+
 proc new*(t: typedesc[HttpSessionRef],
           flags: HttpClientFlags = {},
           maxRedirections = HttpMaxRedirections,
           connectTimeout = HttpConnectTimeout,
           headersTimeout = HttpHeadersTimeout,
           connectionBufferSize = DefaultStreamBufferSize,
-          maxConnections = -1): HttpSessionRef {.
+          maxConnections = -1,
+          idleTimeout = HttpConnectionIdleTimeout,
+          idlePeriod = HttpConnectionCheckPeriod): HttpSessionRef {.
```

Comment:
this is imo an implementation detail that we shouldn't expose in public API - it doesn't matter greatly since we can always make it a no-op, but it feels like the kind of thing that doesn't greatly matter for the user of the API

Reply:
I think some of the users could decrease `idlePeriod` to `1.seconds` just to have more precise control.